apollo-ios/Sources/Apollo/SubscriptableByPathComponent.swift

```diff
@@ -0,0 +1,121 @@
+#if !COCOAPODS
+import ApolloAPI
+#endif
+
+/// Enables subscript syntax to be used where the subscript is an array of `PathComponent` values
+/// typically returned in a GraphQL response.
+protocol SubscriptableByPathComponent {
```

Comment:
This definitely needs a lot of unit test.

I think we should change the name to `PathComponentSubscriptAccessible`

Reply:
We should make the subscript throw and pull the getting and setting out of the extensions on `DataDict` and `[AnyHashable]` to all be in the protocol extension. Checking for correct type for the next path component should be done there.

Reply:
Done in https://github.com/apollographql/apollo-ios-dev/pull/217/commits/dd6b435640e9a376100e3835e16367638b2b0e34.

The refactoring for this includes:
* I got rid of the subscript syntax because the merge process should throw if any errors are found and Swift subscripts cannot throw. There is also no reason it had to be a subscript.
* I pulled all the merging logic into a single file named `DataDict+Merging`. This is the `internal` entry point for getting and setting a value by `[PathComponent]`. The rest of the functions in that file are `fileprivate` because they're internal implementation details. This also negated the need for subscript syntax.
* I kept the protocol but it is renamed to `PathComponentAccessible` and it's needed to escape the type-erased nature of the the protocol for the type-specific mutable setters. We can discuss this in the next review but I think what we've ended up with here is pretty good and concise.